Narrowing default type arguments for D1 type definition

geelen opened the pull request.

petebacondarwin commented on src/cloudflare/internal/d1-api.ts:
```diff
@@ -75,10 +75,10 @@ class D1Database {
   }
 
-  public async exec<T = unknown>(query: string): Promise<D1ExecResult> {
+  public async exec(query: string): Promise<D1ExecResult> {
```
This would be a breaking change to any client who is using `exec()` with a type param, though...

petebacondarwin commented on src/cloudflare/internal/d1-api.ts:
```diff
@@ -276,7 +276,7 @@ class D1PreparedStatement {
   }
 
-  public async raw<T = unknown>(): Promise<T[]> {
+  public async raw<T = unknown[]>(): Promise<T[]> {
```
So, to be clear, the return value is a promise that resolves to an array of arrays, which by default will be `unknown[][]`. Right?

geelen replied:
That's right, yep.

petebacondarwin commented on src/cloudflare/internal/d1-api.ts:
```diff
@@ -266,7 +266,7 @@ class D1PreparedStatement {
   }
 
-  public async all<T = unknown>(): Promise<D1Result<T[]>> {
+  public async all<T = Record<string, unknown>>(): Promise<D1Result<T[]>> {
```
So the default return type here is:

```ts
Promise<D1Result<Record<string, unknown>[]>>
```

So the returned object will have a `results` property which is an array of:

```ts
Record<string, unknown>[]
```

So each row of the `results` will be an array of records. Is that what you want?

petebacondarwin commented on src/cloudflare/internal/d1-api.ts:
```diff
@@ -227,7 +227,7 @@ class D1PreparedStatement {
 
   public async first<T = unknown>(colName: string): Promise<T | null>
-  public async first<T = unknown>(): Promise<Record<string, T> | null>
+  public async first<T = Record<string, unknown>>(): Promise<T | null>
```
OK this looks good. Again this would be a breaking change to any one who has previously used:

```
const result = await statement.first<string>()
```

Since they would have expected a `Record<string, string> | null` but now the type is just `string | null`.
Is that acceptable?

geelen replied:
Yeah that change only _just_ went out, and was an error. You'd rarely have a table of all the same value types, so I think the only people who've tried to use that type are the ones who pointed out the error!

geelen commented:
Thanks for the comments @petebacondarwin @mrbbot, I have just found the `//types:test/types/cf` target that lets me verify the inferred types and actually test-drive this out, so I'll do that tomorrow and resubmit.

geelen commented:
Added a test to cover the definitions, fixed the bug with `.all()`, and I now have a lot more confidence in these changes.

@petebacondarwin and @mrbbot if you wouldn't mind taking a final 👀 at this PR and approving/merging if it all looks good?

geelen commented on types/test/types/d1.ts:
```diff
@@ -0,0 +1,76 @@
+    // FIRST (col)
+    {
+      const result = await stmt.bind(1).first("col1");
+      expectType<unknown | null>(result);
```
Yeah I know, although it wouldn't let me leave off `| null`, so it must be doing _something_. I think it's fine for now, this API will hopefully be supplanted by something with slightly more natural typings in the future, and I'm thinking I'll try to combine the type assertion test with the actual response data test.
